ddtrace/appsec/utils.py

```diff
@@ -13,13 +18,76 @@ def _appsec_rc_features_is_enabled():
 
 def _appsec_rc_capabilities():
     # type: () -> str
-    """return the bit of the composed capabilities in base64
+    r"""return the bit representation of the composed capabilities in base64
     bit 0: Reserved
-    bit 1: ASM Activation
+    bit 1: ASM 1-click Activation
     bit 2: ASM Ip blocking
 
-    TODO: refactor to compose the string and encode it
+    Int Number  -> binary number    -> bytes representation -> base64 representation
+    ASM Activation:
+    2           -> 10               -> b'\x02'              -> "Ag=="
+    ASM Ip blocking:
+    4           -> 100              -> b'\x04'              -> "BA=="
+    ASM Activation and ASM Ip blocking:
+    6           -> 110              -> b'\x06'              -> "Bg=="
+    ...
+    256         -> 100000000        -> b'\x01\x00'          -> b'AQA='
     """
+    value = 0b0
+
     if _appsec_rc_features_is_enabled():
-        return "Ag=="
-    return ""
+        value |= 1 << 1
+        value |= 1 << 2
+
+    result = ""
+    if sys.version_info.major < 3:
+        bytes_res = to_bytes_py2(value, (value.bit_length() + 7) // 8, "big")
+        result = base64.b64encode(bytes_res)  # type: ignore[assignment, arg-type]
+    else:
+        result = str(base64.b64encode(value.to_bytes((value.bit_length() + 7) // 8, "big")), encoding="utf-8")
+
+    return result
+
+
+_HTML_USERDEFINED_BLOCKED_TEMPLATE_CACHE = None  # type: Optional[str]
+_JSON_USERDEFINED_BLOCKED_TEMPLATE_CACHE = None  # type: Optional[str]
+
+
+def _get_blocked_template(accept_header_value):
+    # type: (str) -> str
+
+    global _HTML_USERDEFINED_BLOCKED_TEMPLATE_CACHE
+    global _JSON_USERDEFINED_BLOCKED_TEMPLATE_CACHE
+
+    need_html_template = False
+
+    if accept_header_value and "text/html" in accept_header_value.lower():
+        need_html_template = True
+
+    if need_html_template:
+        template_path = os.getenv("DD_APPSEC_HTTP_BLOCKED_TEMPLATE_HTML")
+    else:
+        template_path = os.getenv("DD_APPSEC_HTTP_BLOCKED_TEMPLATE_JSON")
+
+    if template_path and os.path.exists(template_path) and os.path.isfile(template_path):
+        if need_html_template and _HTML_USERDEFINED_BLOCKED_TEMPLATE_CACHE:
+            return _HTML_USERDEFINED_BLOCKED_TEMPLATE_CACHE
+
+        if not need_html_template and _JSON_USERDEFINED_BLOCKED_TEMPLATE_CACHE:
+            return _JSON_USERDEFINED_BLOCKED_TEMPLATE_CACHE
+        try:
+            with open(template_path, "r") as template_file:
+                content = template_file.read()
+            if need_html_template:
+                _HTML_USERDEFINED_BLOCKED_TEMPLATE_CACHE = content
+            else:
+                _JSON_USERDEFINED_BLOCKED_TEMPLATE_CACHE = content
+            return content
+        except OSError:
+            pass
+
+    # No user-defined template at this point
+    if need_html_template:
+        return constants.APPSEC_BLOCKED_RESPONSE_HTML
+
+    return constants.APPSEC_BLOCKED_RESPONSE_JSON
```

Comment:
They're catched :) (both the user template and the default ones).

Reply:
no, I mean cached as the value of `_JSON_USERDEFINED_BLOCKED_TEMPLATE_CACHE` so we avoid trying to call `with open(...)` on every request if the files are not there.

There is a scenario where the env vars are defined, but the files do not exist, so we will attempt to read the files and get an OSError on every single blocked request.

Is that the expected behavior, or do we want to cache the ultimate end result of this function (obviously html vs json need to be different), so that we avoid any future resolution/logic?

Reply:
I think I understand now, you mean something like "_FILE_NOT_FOUND_PREVIOUSLY" or something like that (will think a better name) and doing:

```       
        except OSError as e:
            log.warning("Could not load custom template at %s: %s", template_path, str(e))  # noqa: G200
            _FILE_NOT_FOUND_PREVIOUSLY = True

```

and then checking it at the start and if `True` avoiding the `open` and directly log the warning (or maybe not even logging the warning after the first one or one in every mod N requests so to not dump too much noise?) right?

Reply:
Found a better way: I also store in the cache the default templates, which are set if the user defined ones are not set or not found, so if `open` fails the default template will be catched and returned and on next requests it will directly return the catched ones without opening.